collector/admission/boundedqueue.go

```diff
@@ -0,0 +1,136 @@
+package admission
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/google/uuid"
+	orderedmap "github.com/wk8/go-ordered-map/v2"
+)
+
+type boundedQueue struct {
+	maxLimitBytes int64
+	maxLimitWaiters int64
+	currentBytes int64
+	currentWaiters int64
+	mtx sync.Mutex
+	// waiters waiters 
+	waiters *orderedmap.OrderedMap[uuid.UUID, waiter]
+}
+
+type waiter struct {
+	readyCh chan struct{} 
+	pendingBytes int64
+	ID uuid.UUID 
+}
+
+func NewBoundedQueue(maxLimitBytes, maxLimitWaiters int64) *boundedQueue {
+	return &boundedQueue{
+		maxLimitBytes: maxLimitBytes,
+		maxLimitWaiters: maxLimitWaiters,
+		currentBytes: int64(0),
+		currentWaiters: int64(0),
+		waiters: orderedmap.New[uuid.UUID, waiter](), 
+	}
+}
+
+func (bq *boundedQueue) admit(pendingBytes int64) (bool, error) {
+	bq.mtx.Lock()
+	defer bq.mtx.Unlock()
+
+	if pendingBytes > bq.maxLimitBytes { // will never succeed
+		return false, fmt.Errorf("rejecting request, request size larger than configured limit")
+	}
+
+	if bq.currentBytes + pendingBytes <= bq.maxLimitBytes { // no need to wait to admit
+		bq.currentBytes += pendingBytes
+		return true, nil
+	}
+
+	// since we were unable to admit, check if we can wait.
+	if bq.currentWaiters + 1 > bq.maxLimitWaiters { // too many waiters
+		return false, fmt.Errorf("rejecting request, too many waiters")
+	}
+
+	// if we got to this point we need to wait to acquire bytes, so update currentWaiters before releasing mutex.
+	bq.currentWaiters += 1
+	return false, nil
+}
+
+func (bq *boundedQueue) Acquire(ctx context.Context, pendingBytes int64) error {
+	success, err := bq.admit(pendingBytes)
+	if err != nil || success {
+		return err
+	}
+
+	// otherwise we need to wait for bytes to be released
+	curWaiter := waiter{
+		pendingBytes: pendingBytes,
+		readyCh: make(chan struct{}),
+		ID: uuid.New(), 
+	}
+
+	bq.mtx.Lock()
+	_, dupped := bq.waiters.Set(curWaiter.ID, curWaiter)
+	if dupped {
+		panic("duplicate keys found")
+	}
```

Comment:
```suggestion
	}

	bq.mtx.Lock()
	for {
	    id := uuid.New()
 	    _, dupped := bq.waiters.Set(id, curWaiter)
	    if dupped {
	        continue
	    }
	    curWaiter.ID = id
	    break
	}
```

Reply:
🤷

Reply:
Thanks for the rewrite, realizing that `bq.waiters.Set(id, curWaiter)` overrides the existing key so I need to use `Get` instead first. Changed to 
```
	bq.lock.Lock()

	for {
	    id := uuid.New()
 	    _, keyExists := bq.waiters.Get(id)
	    if keyExists {
			continue
	    }
	    bq.waiters.Set(id, curWaiter)
            curWaiter.ID = id
            break
	}
```